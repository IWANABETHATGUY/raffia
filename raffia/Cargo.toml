[package]
name = "raffia"
version = "0.1.0"
edition = "2021"
authors = ["Pig Fang <g-plane@hotmail.com>"]
description = "Generic parser for CSS/SCSS/Less."
repository = "https://github.com/g-plane/raffia"
license = "MIT"

[dependencies]
raffia_derive = { path = "../raffia_derive" }
