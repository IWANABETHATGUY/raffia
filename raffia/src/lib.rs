pub use config::Syntax;
pub use parser::{parse, parse_with_comments, Parse, Parser};
pub use pos::{Span, Spanned};

pub mod ast;
mod config;
pub mod error;
mod parser;
pub mod pos;
mod tokenizer;
mod util;
