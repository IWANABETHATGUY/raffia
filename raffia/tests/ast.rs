use insta::{assert_ron_snapshot, glob, Settings};
use raffia::{ast::Stylesheet, Parser, Syntax};
use std::fs;

#[test]
fn ast_snapshot() {
    glob!("ast/**/*.{css,scss,sass,less}", |path| {
        let code = fs::read_to_string(path).unwrap();
        let syntax = match path.extension().unwrap().to_str().unwrap() {
            "css" => Syntax::Css,
            "scss" => Syntax::Scss,
            "sass" => Syntax::Sass,
            "less" => Syntax::Less,
            _ => unreachable!("unknown file extension"),
        };
        let mut parser = Parser::new(&code, syntax);
        let ast = parser.parse::<Stylesheet>().unwrap();

        let mut settings = Settings::clone_current();
        settings.set_snapshot_path(path.parent().unwrap());
        settings.remove_snapshot_suffix();
        settings.set_prepend_module_to_snapshot(false);
        settings.remove_input_file();
        settings.set_omit_expression(true);
        settings.remove_input_file();
        settings.remove_info();
        settings.bind(|| {
            let name = path.file_stem().unwrap().to_str().unwrap();
            assert_ron_snapshot!(name, ast);
        });
    });
}
