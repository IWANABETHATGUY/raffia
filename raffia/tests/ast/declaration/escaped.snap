---
source: raffia/tests/ast.rs
---
Stylesheet(
  type: "Stylesheet",
  statements: [
    QualifiedRule(
      type: "QualifiedRule",
      selector: SelectorList(
        type: "SelectorList",
        selectors: [
          ComplexSelector(
            type: "ComplexSelector",
            children: [
              CompoundSelector(
                type: "CompoundSelector",
                children: [
                  TagNameSelector(
                    type: "TagNameSelector",
                    name: WqName(
                      type: "WqName",
                      name: Ident(
                        type: "Ident",
                        name: "a",
                        raw: "a",
                        span: Span(
                          start: 0,
                          end: 1,
                        ),
                      ),
                      prefix: None,
                      span: Span(
                        start: 0,
                        end: 1,
                      ),
                    ),
                    span: Span(
                      start: 0,
                      end: 1,
                    ),
                  ),
                ],
                span: Span(
                  start: 0,
                  end: 1,
                ),
              ),
            ],
            span: Span(
              start: 0,
              end: 1,
            ),
          ),
        ],
        span: Span(
          start: 0,
          end: 1,
        ),
      ),
      block: SimpleBlock(
        type: "SimpleBlock",
        statements: [
          Declaration(
            type: "Declaration",
            name: Ident(
              type: "Ident",
              name: "color",
              raw: "c\\6flor",
              span: Span(
                start: 6,
                end: 13,
              ),
            ),
            value: [
              HexColor(
                type: "HexColor",
                value: "f00",
                raw: "f00",
                span: Span(
                  start: 15,
                  end: 19,
                ),
              ),
            ],
            important: None,
            lessPropertyMerge: None,
            span: Span(
              start: 6,
              end: 19,
            ),
          ),
          Declaration(
            type: "Declaration",
            name: Ident(
              type: "Ident",
              name: "color",
              raw: "\\63olor",
              span: Span(
                start: 23,
                end: 30,
              ),
            ),
            value: [
              HexColor(
                type: "HexColor",
                value: "f00",
                raw: "f00",
                span: Span(
                  start: 32,
                  end: 36,
                ),
              ),
            ],
            important: None,
            lessPropertyMerge: None,
            span: Span(
              start: 23,
              end: 36,
            ),
          ),
          Declaration(
            type: "Declaration",
            name: Ident(
              type: "Ident",
              name: ",olor",
              raw: "\\2color",
              span: Span(
                start: 40,
                end: 47,
              ),
            ),
            value: [
              HexColor(
                type: "HexColor",
                value: "f00",
                raw: "f00",
                span: Span(
                  start: 49,
                  end: 53,
                ),
              ),
            ],
            important: None,
            lessPropertyMerge: None,
            span: Span(
              start: 40,
              end: 53,
            ),
          ),
          Declaration(
            type: "Declaration",
            name: Ident(
              type: "Ident",
              name: ",olor",
              raw: "\\,olor",
              span: Span(
                start: 57,
                end: 63,
              ),
            ),
            value: [
              HexColor(
                type: "HexColor",
                value: "f00",
                raw: "f00",
                span: Span(
                  start: 65,
                  end: 69,
                ),
              ),
            ],
            important: None,
            lessPropertyMerge: None,
            span: Span(
              start: 57,
              end: 69,
            ),
          ),
        ],
        span: Span(
          start: 2,
          end: 71,
        ),
      ),
      span: Span(
        start: 0,
        end: 71,
      ),
    ),
  ],
  span: Span(
    start: 0,
    end: 72,
  ),
)
